Add Configuration for Higher Time Precision with Decimals

hrmsk66 opened the pull request:
To address https://github.com/davidbarsky/tracing-tree/issues/67, add a new configuration option, `with_higher_precision`. When this option is enabled, the time is displayed with higher precision, including decimal points.

- Added `with_higher_precision` as a configuration option.
- Refactored the existing code in the `on_event` method to separate methods for code reusability[^1]

[^1]: If this PR looks good, I'd like to propose an implementation to address [Issue 17](https://github.com/davidbarsky/tracing-tree/issues/17) based on this as a foundation.

Example Output:
With `higher_precision enabled`, the output will look like this.

```
┐run{hostname="localhost"}
├─ 45.96μs TRACE Request received
├─┐fetch_httpbin{method=GET, path="/"}
│ ├─ 3.70s  DEBUG Response received, backend=httpbin, status=200 OK, cache_status=MISS
├─┘
├─┐fetch_example{method=GET, path="/"}
│ ├─ 587.93ms DEBUG Response received from example, backend=example, status=200 OK, cache_status=HIT
│ ├─┐open_datastore{name="subscriptions"}
│ ├─┘
│ ├─┐read_from_datastore{key="foo@example.com"}
│ │ ├─ 43.04μs  WARN Item found, key=foo@example.com
│ ├─┘
├─┘
├─ 4.29s  TRACE Returning response, status=200 OK
┘
```

oli-obk commented on src/lib.rs:
```diff
@@ -349,39 +427,10 @@ where
 
         // check if this event occurred in the context of a span.
         // if it has, get the start time of this span.
-        let start = match ctx.current_span().id() {
-            Some(id) => match ctx.span(id) {
-                // if the event is in a span, get the span's starting point.
-                Some(ctx) => {
-                    let ext = ctx.extensions();
-                    let data = ext
-                        .get::<Data>()
-                        .expect("Data cannot be found in extensions");
-                    Some(data.start)
-                }
-                None => None,
-            },
-            None => None,
-        };
-        if let Some(start) = start {
-            let elapsed = start.elapsed();
-            let millis = elapsed.as_millis();
-            let secs = elapsed.as_secs();
-            let (n, unit) = if millis < 1000 {
-                (millis as _, "ms")
-            } else if secs < 60 {
-                (secs, "s ")
-            } else {
-                (secs / 60, "m ")
-            };
-            let n = format!("{n:>3}");
-            write!(
-                &mut event_buf,
-                "{timestamp}{unit} ",
-                timestamp = self.styled(Style::new().dimmed(), n),
-                unit = self.styled(Style::new().dimmed(), unit),
-            )
-            .expect("Unable to write to buffer");
```
Tip for the future: when moving things out into a new function, do that in a separate commit, and then modify it in other commits. This makes it easier to track for reviewers to see what changes you made in detail

hrmsk66 replied:
Thanks for the advice. Will separate commits for clarity moving forward.

oli-obk commented on src/lib.rs:
```diff
@@ -291,6 +299,76 @@ where
+    where
+        S: Subscriber + for<'span> LookupSpan<'span>,
+    {
+        match ctx.span(id) {
```
remove the match, now that you got a function where you can use `?`

```suggestion
        let ctx = ctx.span(id)?;
```

hrmsk66 replied:
Fixed as suggested.

oli-obk commented on src/lib.rs:
```diff
@@ -291,6 +299,76 @@ where
+        };
+
+        let n = format!("{n:>3}");
+        format!(
+            "{timestamp}{unit} ",
+            timestamp = self.styled(Style::new().dimmed(), n),
+            unit = self.styled(Style::new().dimmed(), unit),
+        )
```
This is the same in both `format_timestamp*` functions. Maybe move it out of the functions and only return `(n, unit)`

hrmsk66 replied:
Refactored `format_timestamp*` to remove the duplication.

oli-obk reviewed (changes requested):
Just some style nits, then this lgtm

hrmsk66 commented:
Hi @oli-obk 
Rebased to resolve conflicts and noticed my PR needed updates to align with the main branch. In the "concurrent span switching detection (8ccafe8)" commit by @ten3roberts, the steps for retrieving Span's Extensions seemed to have been simplified in the `on_event` method. I've adjusted this PR accordingly.

oli-obk commented:
Thanks!
